diff -r 454ff0629121 account.py
--- a/trytond/trytond/modules/account_invoice/account.py	Tue Nov 08 18:47:58 2016 +0100
+++ b/trytond/trytond/modules/account_invoice/account.py	Fri Dec 02 12:31:52 2016 +0100
@@ -335,4 +335,5 @@
                 ('move', 'in', list(move_ids)),
                 ])
         super(Reconciliation, cls).delete(reconciliations)
-        Invoice.process(invoices)
+        if invoices:
+            Invoice.process(invoices)
