diff -r 8bae00903e21 account.py
--- a/modules/account/account.py    Thu Dec 10 13:03:38 2015 +0100
+++ b/modules/account/account.py    Thu Dec 10 13:06:19 2015 +0100
@@ -1699,11 +1699,20 @@
                     ('account', '=', account_template.id),
                     ('parent', '=', None),
                     ])
-            for tax_template in tax_templates:
-                tax_template.create_tax(self.account.company.id,
+
+            if hasattr(TaxTemplate, 'create_batch'):
+                TaxTemplate.create_batch(
+                    tax_templates,
+                    self.account.company.id,
                     template2tax_code=template2tax_code,
                     template2account=template2account,
                     template2tax=template2tax)
+            else:
+                for tax_template in tax_templates:
+                    tax_template.create_tax(self.account.company.id,
+                        template2tax_code=template2tax_code,
+                        template2account=template2account,
+                        template2tax=template2tax)
 
             # Update taxes on accounts
             account_template.update_account_taxes(template2account,
@@ -1726,6 +1735,7 @@
             for tax_rule_line_template in tax_rule_line_templates:
                 tax_rule_line_template.create_rule_line(template2tax,
                     template2rule, template2rule_line=template2rule_line)
+
         return 'properties'
 
     def default_properties(self, fields):

diff -r 21688b4ca06a tryton.cfg
--- a/modules/account_es/tryton.cfg	Tue Aug 11 16:16:46 2015 +0200
+++ b/modules/account_es/tryton.cfg	Thu Sep 03 16:48:10 2015 +0200
@@ -3,6 +3,7 @@
 depends:
     ir
     account
+    account_chart_speedup
 xml:
     account.xml
     view.xml
