diff -r 74bdc279fae0 asset.py
--- a/trytond/trytond/modules/account_asset/asset.py	Mon May 06 14:53:17 2019 +0200
+++ b/trytond/trytond/modules/account_asset/asset.py	Tue Jun 11 22:56:45 2019 +0200
@@ -922,6 +922,8 @@
             def start_fixed_value(self):
                 if self.start_date < self.asset.start_date:
                     return 0
+                if not self.asset_lines:
+                    return 0
                 value = self.asset_lines[0].acquired_value
                 date = self.asset_lines[0].date
                 for line in self.update_lines:
