diff -r b932cbd06f9f invoice.py
--- a/trytond/trytond/modules/commmision/invoice.py	Mon May 21 20:07:09 2018 +0200
+++ b/trytond/trytond/modules/commission/invoice.py	Mon Aug 27 10:24:36 2018 +0200
@@ -130,7 +130,6 @@
             'invisible': ~Eval('from_commissions'),
             })
 
-    @property
     def agent_plans_used(self):
         "List of agent, plan tuple"
         used = []
@@ -151,7 +150,7 @@
 
         today = Date.today()
         commissions = []
-        for agent, plan in self.agent_plans_used:
+        for agent, plan in self.agent_plans_used():
             if not plan:
                 continue
             with Transaction().set_context(date=self.invoice.currency_date):
