diff -r 2ae55272d6e6 move.py
--- a/modules/stock/move.py	Thu Jun 15 12:27:08 2017 +0200
+++ b/modules/stock/move.py	Thu Jun 15 15:31:27 2017 +0200
@@ -625,7 +625,6 @@
         for move in moves:
             move.set_effective_date()
             to_write.extend(([move], {
-                        'effective_date': move.effective_date,
                         'state': 'assigned'
                         }))
         if to_write:
@@ -892,7 +891,8 @@
             stock_date_start: if set return the delta of the stock between the
                 two dates, (ignored if stock_date_end is missing).
             stock_assign: if set compute also the assigned outgoing moves as
-                done.
+                done at the stock_date_end except for delta which is at the
+                planned date.
             forecast: if set compute the forecast quantity.
             stock_destinations: A list of location ids. If set, restrict the
                 computation to moves from and to those locations.
@@ -958,6 +958,9 @@
                         & (move.planned_date <= context['stock_date_end'])
                         )
                     | (move.effective_date <= context['stock_date_end'])
+                    | (move.state == (
+                            'assigned' if not context.get('stock_date_start')
+                            else ''))
                     )
                 )
             state_date_clause_in = state_date_clause(False)
@@ -976,6 +979,10 @@
                             & (move.planned_date <= today)
                             )
                         | (move.effective_date <= today)
+                        | (move.state == (
+                            'assigned'
+                            if not context.get('stock_date_start')
+                            else ''))
                         )
                     )
                 | (move.state.in_(['done', 'assigned', 'draft'])
@@ -991,6 +998,10 @@
                             (move.effective_date <= context['stock_date_end'])
                             & (move.effective_date >= today)
                             )
+                        | (move.state == (
+                            'assigned'
+                            if not context.get('stock_date_start')
+                            else ''))
                         )
                     )
                 )
