diff -r a87800ef903a move.py
--- a/trytond/trytond/modules/stock/move.py	Fri Nov 27 12:13:17 2015 +0100
+++ b/trytond/trytond/modules/stock/move.py	Fri Nov 27 12:13:55 2015 +0100
@@ -775,7 +775,7 @@
         Date = pool.get('ir.date')
         Location = pool.get('stock.location')
 
-        Transaction().cursor.lock(cls._table)
+        # Transaction().cursor.lock(cls._table)
 
         if with_childs:
             locations = Location.search([
