diff -r e0e13043926b template.py
--- ./modules/electronic_mail_template/template.py	Mon Feb 17 16:06:53 2014 +0100
+++ ./modules/electronic_mail_template/template.py	Wed Mar 19 11:04:42 2014 +0100
@@ -353,7 +353,7 @@
 
         """Validate recipients to send or move email to draft mailbox"""
         emails = ",".join(recepients)
-        if not ElectronicMail.get_email_valid(emails) and template:
+        if not ElectronicMail.validate(emails) and template:
             """Draft Mailbox. Not send email"""
             ElectronicMail.write([email], {
                 'mailbox': template.draft_mailbox,
