diff --git a/trytond/ir/action.py b/trytond/ir/action.py
--- a/trytond/trytond/ir/action.py
+++ b/trytond/trytond/ir/action.py
@@ -149,13 +149,13 @@
     def check_wizard_model(self):
         ActionWizard = Pool().get('ir.action.wizard')
         if self.action.type == 'ir.action.wizard':
-            action_wizard, = ActionWizard.search([
+            action_wizards = ActionWizard.search([
                 ('action', '=', self.action.id),
                 ], limit=1)
-            if action_wizard.model:
-                if self.model.__name__ != action_wizard.model:
+            if action_wizards and action_wizards[0].model:
+                if self.model.__name__ != action_wizards[0].model:
                     self.raise_user_error('wrong_wizard_model', (
-                            action_wizard.rec_name,))
+                            action_wizards[0].rec_name,))
 
     @staticmethod
     def _convert_vals(vals):
