diff -r 02a162618a67 trytond/trytond/modules/sale/sale.py
--- a/trytond/trytond/modules/sale/sale.py	Wed Oct 25 18:48:58 2017 +0200
+++ b/trytond/trytond/modules/sale/sale.py	Fri Nov 10 13:28:49 2017 +0100
@@ -928,7 +928,7 @@
                     if sale.state == 'confirmed':
                         process.append(sale)
                     done.append(sale)
-            elif sale.state != 'processing':
+            if sale.state != 'processing':
                 process.append(sale)
         if process:
             cls.proceed(process)
