diff -r 254e4bb38458 carrier.py
--- a/trytond/trytond/modules/carrier/carrier.py	Mon May 06 15:01:12 2019 +0200
+++ b/trytond/trytond/modules/carrier/carrier.py	Mon Jun 10 10:00:08 2019 +0200
@@ -13,6 +13,8 @@
 class Carrier(ModelSQL, ModelView):
     'Carrier'
     __name__ = 'carrier'
+    _rec_name = 'party'
+
     party = fields.Many2One('party.party', 'Party', required=True,
             ondelete='CASCADE', help="The party which represents the carrier.")
     carrier_product = fields.Many2One('product.product', 'Carrier Product',
