diff -r a6a39965d1f8 trytond/trytond/modules/country/scripts/import_zip.py
--- a/trytond/trytond/modules/country/scripts/import_zip.py	Mon Nov 04 12:58:46 2019 +0100
+++ b/trytond/trytond/modules/country/scripts/import_zip.py	Thu Nov 21 21:22:22 2019 +0100
@@ -81,7 +81,7 @@
     for row in pbar(list(csv.DictReader(
                     f, fieldnames=_fieldnames, delimiter='\t'))):
         country = get_country(row['country'])
-        subdivision = get_subdivision(row['country'], row['code1'])
+        subdivision = get_subdivision(row['country'], row['code2'])
         zips.append(
             Zip(country=country, subdivision=subdivision, zip=row['postal'],
             city=row['place']))
