diff -r c7607908611c trytond/trytond/modules/account_invoice/invoice.py
--- a/trytond/trytond/modules/account_invoice/invoice.py	Wed May 11 11:37:56 2016 +0200
+++ b/trytond/trytond/modules/account_invoice/invoice.py	Mon May 30 15:28:43 2016 +0200
@@ -1018,9 +1018,12 @@
                 cls.raise_user_error('modify_invoice', (invoice.rec_name,))
 
     def get_rec_name(self, name):
-        return (self.number or unicode(self.id)
-            + (self.reference and (' ' + self.reference) or '')
-            + ' ' + self.party.rec_name)
+        name = self.number or unicode(self.id)
+        if self.reference:
+            name += ' %s' % self.reference
+        if self.party:
+            name += ' %s' % self.party.rec_name
+        return name
 
     @classmethod
     def search_rec_name(cls, name, clause):
