Index: tryton/tryton/gui/window/view_form/view/list_gtk/widget.py
===================================================================

--- a/tryton/tryton/gui/window/view_form/view/list_gtk/widget.py
+++ b/tryton/tryton/gui/window/view_form/view/list_gtk/widget.py
@@ -809,6 +809,7 @@

     def __init__(self, view, attrs):
         super(ProgressBar, self).__init__()
+        self.view = view
         self.attrs = attrs
         self.renderer = gtk.CellRendererProgress()
         orientation = self.orientations.get(self.attrs.get('orientation',

