diff -r c5fad88f4be1 stock.py
--- a/trytond/trytond/modules/stock_consignment/stock.py	Mon Oct 30 15:55:27 2017 +0100
+++ b/trytond/trytond/modules/stock_consignment/stock.py	Fri Dec 15 02:05:38 2017 +0100
@@ -107,7 +107,7 @@
 
     def get_invoice_line_consignment(self):
         if (self.from_location.type == 'supplier'
-                and self.to_location.type == 'storage'
+                and self.to_location.type in ('storage', 'production')
                 and self.from_location.consignment_party):
             return self._get_supplier_invoice_line_consignment()
         elif (self.from_location.type == 'storage'
