diff -r 211d4069243a trytond/trytond/modules/stock_supply_production/production.py
--- a/trytond/trytond/modules/stock_supply_production/production.py	Thu Jan 31 12:27:10 2019 +0100
+++ b/trytond/trytond/modules/stock_supply_production/production.py	Thu Jan 31 12:27:58 2019 +0100
@@ -130,7 +130,7 @@
             product=product,
             bom=product.boms[0].bom if product.boms else None,
             uom=product.default_uom,
-            quantity=quantity,
+            quantity=float(Decimal(quantity).quantize(Decimal(str(10 ** -product.default_uom.digits)))),
             state='request',
             )
