Index: move.py
===================================================================

--- ./modules/account/move.py
+++ ./modules/account/move.py
@@ -358,8 +358,8 @@
             cls.write([move], values)
 
             to_reconcile = [l for l in move.lines
-                if ((line.debit == line.credit == Decimal('0'))
-                    and line.account.reconcile)]
+                if ((l.debit == l.credit == Decimal('0'))
+                    and l.account.reconcile)]
             if to_reconcile:
                 Line.reconcile(to_reconcile)
 

