Index: tryton/gui/window/view_form/model/field.py
===================================================================

--- ./tryton/tryton/gui/window/view_form/model/field.py
+++ ./tryton/tryton/gui/window/view_form/model/field.py
@@ -253,7 +253,7 @@
     _default = None
 
     def set_client(self, record, value, force_change=False):
-        if not isinstance(value, datetime.time) and value is not None:
+        if isinstance(value, basestring) and value is not None:
             try:
                 value = datetime.time(*time.strptime(value,
                         self.time_format(record))[3:6])

