Index: trytond/trytond/ir/ui/tree.rnc
===================================================================

--- a/trytond/trytond/ir/ui/tree.rnc
+++ b/trytond/trytond/ir/ui/tree.rnc
@@ -39,6 +39,7 @@
     | "reference"
     | "one2one"
     | "binary"
+    | "image"
   }?
 attlist.field &=
   [ a:defaultValue = "0" ] attribute tree_invisible { "0" | "1" }?
@@ -47,6 +48,7 @@
 attlist.field &= attribute icon { text }?
 attlist.field &= attribute sum { text }?
 attlist.field &= attribute width { text }?
+attlist.field &= attribute height { text }?
 attlist.field &=
   [ a:defaultValue = "left_to_right" ] attribute orientation {
     "left_to_right"

Index: trytond/trytond/ir/ui/tree.rng
===================================================================

--- a/trytond/trytond/ir/ui/tree.rng
+++ b/trytond/trytond/ir/ui/tree.rng
@@ -101,6 +101,7 @@
           <value>reference</value>
           <value>one2one</value>
           <value>binary</value>
+          <value>image</value>
         </choice>
       </attribute>
     </optional>
@@ -142,6 +143,11 @@
   </define>
   <define name="attlist.field" combine="interleave">
     <optional>
+      <attribute name="height"/>
+    </optional>
+  </define>
+  <define name="attlist.field" combine="interleave">
+    <optional>
       <attribute name="orientation" a:defaultValue="left_to_right">
         <choice>
           <value>left_to_right</value>

