Index: trytond/model/modelview.py
===================================================================

--- a/trytond/trytond/model/modelview.py
+++ b/trytond/trytond/model/modelview.py
@@ -426,7 +426,9 @@
         # convert attributes into pyson
         encoder = PYSONEncoder()
         for attr in ('states', 'domain', 'spell', 'colors'):
-            if element.get(attr):
+            if (element.get(attr)
+                    # Avoid double evaluation from inherit with different model
+                    and '__' not in element.get(attr)):
                 element.set(attr, encoder.encode(safe_eval(element.get(attr),
                     CONTEXT)))


