Index: tryton/tryton/gui/window/view_form/view/list_gtk/widget.py
===================================================================

--- a/tryton/tryton/gui/window/view_form/view/list_gtk/widget.py
+++ b/tryton/tryton/gui/window/view_form/view/list_gtk/widget.py
@@ -530,7 +530,8 @@
             else:
                 value = field.get_data(record)
         pixbuf = data2pixbuf(value)
-        if self.attrs['width'] != -1 or self.attrs['height'] != -1:
+        if (self.attrs.get('width', -1) != -1 or
+                self.attrs.get('height', -1) != -1):
             pixbuf = common.resize_pixbuf(pixbuf,
                 self.attrs['width'], self.attrs['height'])
         cell.set_property('pixbuf', pixbuf)

