diff -r b74f331bafc9 trytond/trytond/modules/sale_supply_drop_shipment/stock.py
--- a/trytond/trytond/modules/sale_supply_drop_shipment/stock.py	Sun Oct 02 15:21:28 2016 +0200
+++ b/trytond/trytond/modules/sale_supply_drop_shipment/stock.py	Thu Oct 06 08:20:31 2016 +0200
@@ -399,7 +399,7 @@
 
         requests = []
         for sub_lines in grouped_slice([m.origin.id for s in shipments
-                    for m in s.supplier_moves]):
+                    for m in s.supplier_moves if m.origin]):
             requests += PurchaseRequest.search([
                     ('purchase_line', 'in', list(sub_lines)),
                     ])
@@ -412,6 +412,8 @@
         to_save = []
         for shipment in shipments:
             for move in shipment.supplier_moves:
+                if not move.origin:
+                    continue
                 sale_line = request2sline[pline2request[move.origin]]
                 for move in sale_line.moves:
                     if (move.state not in ('cancel', 'done')
