Index: configuration.py
===================================================================

--- a/trytod/trytond/modules/product/configuration.py
+++ b/trytod/trytond/modules/product/configuration.py
@@ -27,6 +27,12 @@
     default_cost_price_method = fields.MultiValue(default_cost_price_method)
     get_cost_price_methods = get_cost_price_methods

+    def get_multivalue(self, name, **pattern):
+        if 'company' in pattern:
+            pattern = pattern.copy()
+            pattern.pop('company')
+        return super(Configuration, self).get_multivalue(name, **pattern)
+
     @classmethod
     def default_default_cost_price_method(cls, **pattern):
         return cls.multivalue_model(
