diff -r 365edd0bc937 trytond/trytond/model/modelstorage.py
--- a/trytond/trytond/model/modelstorage.py	Wed Jul 11 15:15:17 2018 +0200
+++ b/trytond/trytond/model/modelstorage.py	Wed Jul 11 15:17:53 2018 +0200
@@ -1006,10 +1006,12 @@
             if relations:
                 for sub_relations in grouped_slice(relations):
                     sub_relations = set(sub_relations)
-                    finds = Relation.search(['AND',
-                            [('id', 'in', [r.id for r in sub_relations])],
-                            domain,
-                            ])
+                    # Use root user to skip access rules
+                    with Transaction().set_user(0):
+                        finds = Relation.search(['AND',
+                                [('id', 'in', [r.id for r in sub_relations])],
+                                domain,
+                                ])
                     if sub_relations != set(finds):
                         fail_field_value = sub_relations.difference(
                             set(finds)).pop()
