Index: trytond/trytond/model/fields/one2one.py
===================================================================

--- a/trytond/trytond/model/fields/one2one.py
+++ b/trytond/rytond/model/fields/one2one.py
@@ -25,7 +25,7 @@
         '''
         res = super(One2One, self).get(ids, model, name, values=values)
         for i, vals in res.iteritems():
-            res[i] = vals[0] if vals else False
+            res[i] = vals[0] if vals else None
         return res

     def set(self, Model, name, ids, value, *args):

