diff -r f91027954501 trytond/trytond/model/multivalue.py
--- a/trytond/trytond/model/multivalue.py	Mon Jun 11 23:17:54 2018 +0200
+++ b/trytond/trytond/model/multivalue.py	Thu Jun 28 09:59:05 2018 +0200
@@ -34,8 +34,10 @@
         return Value(**pattern)

     def __values(self, field, pattern, match_none=True):
-        return [v for v in self.multivalue_records(field)
-            if v.match(pattern, match_none=match_none)]
+        Value = self.multivalue_model(field)
+        # Limit the cache to matching records
+        return Value.browse((v for v in self.multivalue_records(field)
+            if v.match(pattern, match_none=match_none)))

     def get_multivalue(self, name, **pattern):
         Value = self.multivalue_model(name)
