Index: ./trytond/trytond/modules/stock_supply_production/production.py
===================================================================
--- ./trytond/trytond/modules/stock_supply_production/production.py
+++ ./trytond/trytond/modules/stock_supply_production/production.py
@@ -1,6 +1,7 @@
 # This file is part of Tryton.  The COPYRIGHT file at the top level of
 # this repository contains the full copyright notices and license terms.
 import datetime
+from decimal import Decimal
 from collections import defaultdict
 
 from trytond.model import ModelSQL, ValueMixin, fields
@@ -141,7 +142,7 @@ class Production(metaclass=PoolMeta):
             product=product,
             bom=product.boms[0].bom if product.boms else None,
             uom=product.default_uom,
-            quantity=quantity,
+            quantity=float(Decimal(quantity).quantize(Decimal(str(10 ** -product.default_uom.digits)))),
             state='request',
             )
 
