Index: tryton/tryton/gui/window/view_form/model/field.py
===================================================================

--- a/tryton/tryton/gui/window/view_form/model/field.py
+++ b/tryton/tryton/gui/window/view_form/model/field.py
@@ -135,7 +135,8 @@
         previous_value = self.get(record)
         self.set(record, value)
         if previous_value != self.get(record):
-            record.modified_fields.setdefault(self.name)
+            if not self.attrs.get('readonly'):
+                record.modified_fields.setdefault(self.name)
             record.signal('record-modified')
             self.sig_changed(record)
             record.validate(softvalidation=True)

