Index: payment_term.py
===================================================================
--- a/trytond/trytond/modules/account_invoice/payment_term.py
+++ b/trytond/trytond/modules/account_invoice/payment_term.py
@@ -222,11 +222,10 @@
                     gettext('account_invoice'
                         '.msg_payment_term_invalid_ratio_divisor',
                         line=line.rec_name))
-            ratio = line.ratio
-            divisor = line.divisor
-            line.on_change_ratio()
-            line.on_change_divisor()
-            if (line.divisor != divisor) or (line.ratio != ratio):
+            if (line.ratio != round(
+                        1 / line.divisor, cls.ratio.digits[1])
+                    and line.divisor != round(
+                        1 / line.ratio, cls.divisor.digits[1])):
                 raise PaymentTermValidationError(
                     gettext('account_invoice'
                         '.msg_payment_term_invalid_ratio_divisor',
