Index: trytond/trytond/modules/purchase/product.py
===================================================================
--- a/trytond/trytond/modules/purchase/product.py
+++ b/trytond/trytond/modules/purchase/product.py
@@ -91,6 +91,15 @@ class Template(metaclass=PoolMeta):
                             name, gettext('purchase.msg_change_purchase_uom'))
         super(Template, cls).write(*args)
 
+    @classmethod
+    def copy(cls, templates, default=None):
+        if default is None:
+            default = {}
+        else:
+            default = default.copy()
+        default.setdefault('product_suppliers', None)
+        return super(Template, cls).copy(templates, default=default)
+
 
 class Product(metaclass=PoolMeta):
     __name__ = 'product.product'
@@ -110,6 +119,15 @@ class Product(metaclass=PoolMeta):
             "Purchase Price", digits=price_digits), 'get_purchase_price_uom')
 
     @classmethod
+    def copy(cls, products, default=None):
+        if default is None:
+            default = {}
+        else:
+            default = default.copy()
+        default.setdefault('product_suppliers', None)
+        return super(Product, cls).copy(products, default=default)
+
+    @classmethod
     def get_purchase_price_uom(cls, products, name):
         quantity = Transaction().context.get('quantity') or 0
         return cls.get_purchase_price(products, quantity=quantity)
