diff -r 90cc3f4c20c9 trytond/trytond/model/modelstorage.py
--- a/trytond/trytond/model/modelstorage.py	Tue Sep 22 16:39:45 2015 +0200
+++ b/trytond/trytond/model/modelstorage.py	Tue Sep 22 16:40:32 2015 +0200
@@ -952,10 +952,11 @@
             if relations:
                 for sub_relations in grouped_slice(relations):
                     sub_relations = set(sub_relations)
-                    finds = Relation.search(['AND',
-                            [('id', 'in', [r.id for r in sub_relations])],
-                            domain,
-                            ])
+                    with Transaction().set_user(0):
+                        finds = Relation.search(['AND',
+                                [('id', 'in', [r.id for r in sub_relations])],
+                                domain,
+                                ])
                     if sub_relations != set(finds):
                         cls.raise_user_error('domain_validation_record',
                             error_args=cls._get_error_args(field.name))
