diff -r 90cc3f4c20c9 trytond/trytond/model/modelstorage.py
--- a/trytond/trytond/model/modelstorage.py	Tue Sep 22 16:39:45 2015 +0200
+++ b/trytond/trytond/model/modelstorage.py	Tue Sep 22 16:40:32 2015 +0200
@@ -928,10 +928,11 @@
             else:
                 relations = records
             if relations:
-                finds = Relation.search(['AND',
-                        [('id', 'in', [r.id for r in relations])],
-                        domain,
-                        ])
+                with Transaction().set_user(0):
+                    finds = Relation.search(['AND',
+                            [('id', 'in', [r.id for r in relations])],
+                            domain,
+                            ])
                 if set(relations) != set(finds):
                     cls.raise_user_error('domain_validation_record',
                         error_args=cls._get_error_args(field.name))
