diff -r f4710591683e move.py
--- a/trytond/trytond/modules/product_cost_fifo/move.py	Mon May 06 15:06:16 2019 +0200
+++ b/trytond/trytond/modules/product_cost_fifo/move.py	Wed Oct 02 10:57:59 2019 +0200
@@ -70,6 +70,9 @@

         if Decimal(str(consumed_qty)) != Decimal("0"):
             cost_price = cost_price / Decimal(str(consumed_qty))
+        else:
+            cost_price = self.product.get_multivalue(
+                'cost_price', **self._cost_price_pattern)

         # Compute average cost price
         unit_price = self.unit_price
