Index: invoice.py
===================================================================

--- ./modules/account_invoice/invoice.py
+++ ./modules/account_invoice/invoice.py
@@ -1504,7 +1504,7 @@
         on_change=['product', 'unit', 'quantity', 'description',
             '_parent_invoice.type', '_parent_invoice.party',
             '_parent_invoice.currency', '_parent_invoice.currency_date',
-            'party', 'currency', 'invoice'],
+            'party', 'currency', 'invoice', 'invoice_type'],
         depends=['type'])
     product_uom_category = fields.Function(
         fields.Many2One('product.uom.category', 'Product Uom Category',

