Index: tryton/tryton/gui/window/view_form/model/field.py
===================================================================

--- a/tryton/tryton/gui/window/view_form/model/field.py
+++ b/tryton/tryton/gui/window/view_form/model/field.py
@@ -671,6 +671,7 @@
         record.modified_fields.setdefault(self.name)

     def set_on_change(self, record, value):
+        self._set_default_value(record)
         if isinstance(value, (list, tuple)):
             self._set_value(record, value)
             record.modified_fields.setdefault(self.name)

