diff -r af895bf647f2 product.py
--- a/trytond/trytond/modules/product_cost_fifo/product.py	Fri Aug 09 16:36:11 2019 +0200
+++ b/trytond/trytond/modules/product_cost_fifo/product.py	Thu Sep 19 00:59:26 2019 +0200
@@ -50,7 +50,8 @@
             moves = Move.search([
                 ('product', '=', product.id),
                 ('state', '=', 'done'),
-                ('from_location.type', '=', 'supplier'),
+                ('from_location.type', 'in', ['supplier', 'production']),
+                ('to_location.type', '=', 'storage'),
                 ], offset=offset, limit=limit,
                 order=[('effective_date', 'DESC'), ('id', 'DESC')])
             if not moves:
