Index: tryton/tryton/common/common.py
===================================================================

--- a/tryton/tryton/common/common.py
+++ b/tryton/tryton/common/common.py
@@ -1577,8 +1577,10 @@


 def data2pixbuf(data):
-    try:
-        pixbuf = _data2pixbuf(data)
-    except glib.GError:
-        pixbuf = NO_IMG_PIXBUF
+    pixbuf = NO_IMG_PIXBUF
+    if data:
+        try:
+            pixbuf = _data2pixbuf(data)
+        except glib.GError:
+            pass
     return pixbuf

