diff -r 9b67a6fb35c2 trytond/trytond/ir/ui/tree.rnc
--- a/trytond/trytond/ir/ui/tree.rnc	Sun Oct 06 13:45:55 2019 +0200
+++ b/trytond/trytond/ir/ui/tree.rnc	Thu Oct 10 15:00:17 2019 +0200
@@ -13,7 +13,7 @@
   [ a:defaultValue = "0" ] attribute tree_state { "0" | "1" }?
 field = element field { attlist.field, (prefix | suffix)* }
 attlist.field &= attribute name { text }
-attlist.field &= attribute readonly { "0" | "1" }?
+[ a:defaultValue = "0" ] attribute expand { xsd:integer }?
 attlist.field &=
   attribute widget {
     "char"
diff -r 9b67a6fb35c2 trytond/trytond/ir/ui/tree.rng
--- a/trytond/trytond/ir/ui/tree.rng	Sun Oct 06 13:45:55 2019 +0200
+++ b/trytond/trytond/ir/ui/tree.rng	Thu Oct 10 15:00:17 2019 +0200
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <grammar xmlns="http://relaxng.org/ns/structure/1.0"
-         xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0">
+         xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0"
+         datatypeLibrary="http://www.w3.org/2001/XMLSchema-datatypes">
   <define name="tree">
     <element>
       <name ns="">tree</name>
@@ -135,10 +136,7 @@
     <optional>
       <attribute a:defaultValue="0">
         <name ns="">expand</name>
-        <choice>
-          <value>0</value>
-          <value>1</value>
-        </choice>
+        <data type="integer"/>
       </attribute>
     </optional>
   </define>
