diff -r 6c0d835dd918 trytond/trytond/model/modelstorage.py
--- a/trytond/trytond/model/modelstorage.py	Wed May 11 10:40:32 2016 +0200
+++ b/trytond/trytond/model/modelstorage.py	Mon May 30 15:46:22 2016 +0200
@@ -954,10 +954,11 @@
             if relations:
                 for sub_relations in grouped_slice(relations):
                     sub_relations = set(sub_relations)
-                    finds = Relation.search(['AND',
-                            [('id', 'in', [r.id for r in sub_relations])],
-                            domain,
-                            ])
+                    with Transaction().set_user(0):
+                        finds = Relation.search(['AND',
+                                [('id', 'in', [r.id for r in sub_relations])],
+                                domain,
+                                ])
                     if sub_relations != set(finds):
                         cls.raise_user_error('domain_validation_record',
                             error_args=cls._get_error_args(field.name))
