Index: trytond/trytond/modules/sale/sale.py
===================================================================

--- a/trytond/trytond/modules/sale/sale.py
+++ b/trytond/trytond/modules/sale/sale.py
@@ -1347,7 +1347,8 @@
                     old_invoice_line.quantity, self.unit)
         invoice_line.quantity = quantity
 
-        if invoice_line.quantity <= 0.0:
+        rounding = self.unit.rounding if self.unit else 0
+        if invoice_line.quantity < rounding:
             return []
         invoice_line.unit = self.unit
         invoice_line.product = self.product
@@ -1418,7 +1419,7 @@
             if move.id not in skip_ids:
                 quantity -= Uom.compute_qty(move.uom, move.quantity,
                     self.unit)
-        if quantity <= 0.0:
+        if quantity < self.unit.rounding:
             return
         if not self.sale.party.customer_location:
             self.raise_user_error('customer_location_required', {

