Index: imi/trytond/trytond/modules/sale_supply_drop_shipment/stock.py
===================================================================
--- imi.orig/trytond/trytond/modules/sale_supply_drop_shipment/stock.py
+++ imi/trytond/trytond/modules/sale_supply_drop_shipment/stock.py
@@ -522,7 +522,10 @@ class ShipmentDrop(Workflow, ModelSQL, M
             for move in shipment.supplier_moves:
                 if not move.origin:
                     continue
-                sale_line = request2sline.get(pline2request.get(move.origin))
+                purchase_request = pline2request.get(move.origin)
+                if not purchase_request:
+                    continue
+                sale_line = request2sline[purchase_request]
                 if not sale_line:
                     continue
                 for move in sale_line.moves:
