diff -r 758b49de71a9 trytond/trytond/modules/sale/sale.py
--- a/trytond/trytond/modules/sale/sale.py	Wed May 11 11:54:40 2016 +0200
+++ b/trytond/trytond/modules/sale/sale.py	Mon May 30 15:22:18 2016 +0200
@@ -919,7 +919,7 @@
             if sale.is_done():
                 if sale.state != 'done':
                     done.append(sale)
-            elif sale.state != 'processing':
+            if sale.state != 'processing':
                 process.append(sale)
         if process:
             cls.proceed(process)
