Index: purchase_request.py
===================================================================

--- ./modules/stock_supply/purchase_request.py
+++ ./modules/stock_supply/purchase_request.py
@@ -677,7 +677,7 @@
                     setattr(purchase, f, v)
                 purchase.save()
                 for request in grouped_requests:
-                    line = self.compute_purchase_line(request)
+                    line = self.compute_purchase_line(request, purchase)
                     line.purchase = purchase
                     line.save()
                     Request.write([request], {
@@ -693,7 +693,7 @@
         return {}
 
     @classmethod
-    def compute_purchase_line(cls, request):
+    def compute_purchase_line(cls, request, purchase):
         pool = Pool()
         Product = pool.get('product.product')
         Line = pool.get('purchase.line')

