Index: trytond/trytond/modules/production/bom.py
===================================================================

--- a/trytond/trytond/modules/production/bom.py
+++ b/trytond/trytond/modules/production/bom.py
@@ -67,6 +67,7 @@
     unit_digits = fields.Function(fields.Integer('Unit Digits'),
         'on_change_with_unit_digits')
     quantity = fields.Float('Quantity', required=True,
+        domain=[('quantity', '>', 0)],
         digits=(16, Eval('unit_digits', 2)),
         depends=['unit_digits'])


