diff --git a/trytond/model/fields/many2one.py b/trytond/model/fields/many2one.py
index 62344269..529e91b1 100644
--- a/trytond/trytond/model/fields/many2one.py
+++ b/trytond/model/fields/many2one.py
@@ -26,7 +26,7 @@ class Many2One(Field):
             search_order=None, search_context=None, help='', required=False,
             readonly=False, domain=None, states=None, select=False,
             on_change=None, on_change_with=None, depends=None, context=None,
-            loading='eager'):
+            loading='lazy'):
         '''
         :param model_name: The name of the target model.
         :param left: The name of the field to store the left value for
diff --git a/trytond/model/modelstorage.py b/trytond/model/modelstorage.py
index a6141be9..9ee6caba 100644
--- a/trytond/trytond/model/modelstorage.py
+++ b/trytond/trytond/model/modelstorage.py
@@ -99,11 +99,11 @@ class ModelStorage(Model):
     create_uid = fields.Many2One(
         'res.user', lazy_gettext('ir.msg_created_by'), readonly=True)
     create_date = fields.Timestamp(
-        lazy_gettext('ir.msg_created_by'), readonly=True)
+        lazy_gettext('ir.msg_created_by'), readonly=True, loading='lazy')
     write_uid = fields.Many2One(
         'res.user', lazy_gettext('ir.msg_edited_by'), readonly=True)
     write_date = fields.Timestamp(
-        lazy_gettext('ir.msg_edited_at'), readonly=True)
+        lazy_gettext('ir.msg_edited_at'), readonly=True, loading='lazy')
     rec_name = fields.Function(
         fields.Char(lazy_gettext('ir.msg_record_name')), 'get_rec_name',
         searcher='search_rec_name')
