diff -r 26a124068f61 res.py
--- a/trytond/trytond/modules/ldap_authentication/res.py	Mon Apr 21 21:00:24 2014 +0200
+++ b/trytond/trytond/modules/ldap_authentication/res.py	Fri Jul 11 16:03:59 2014 +0200
@@ -131,7 +131,7 @@
             [(dn, attrs)] = cls.ldap_search_user(login, con, connection,
                 attrs=[str(connection.auth_uid)])
             if password and con.simple_bind_s(dn, password):
-                user_id, _, _ = cls._get_login(login)
+                user_id, _ = cls._get_login(login)
                 if user_id:
                     LoginAttempt.remove(login)
                     return user_id
