diff -r 9fca89e5d103 trytond/trytond/modules/stock/move.py
--- a/trytond/trytond/modules/stock/move.py	Tue Sep 20 16:35:04 2016 +0200
+++ b/trytond/trytond/modules/stock/move.py	Tue Sep 20 16:36:31 2016 +0200
@@ -746,6 +746,10 @@
             return to_pick
         return to_pick
 
+    @staticmethod
+    def lock_stock_move():
+        return True
+
     @classmethod
     def assign_try(cls, moves, with_childs=True, grouping=('product',)):
         '''
@@ -759,7 +763,8 @@
         Date = pool.get('ir.date')
         Location = pool.get('stock.location')
 
-        Transaction().cursor.lock(cls._table)
+        if cls.lock_stock_move():
+            Transaction().cursor.lock(cls._table)
 
         if with_childs:
             locations = Location.search([
