diff -r 7ed03689fe01 trytond/trytond/protocols/dispatcher.py
--- a/trytond/trytond/protocols/dispatcher.py	Fri May 11 13:52:50 2018 +0200
+++ b/trytond/trytond/protocols/dispatcher.py	Fri May 11 13:56:27 2018 +0200
@@ -192,7 +192,7 @@
                 raise
             except (ConcurrencyException, UserError, UserWarning,
                     LoginException):
-                logger.debug(log_message, *log_args, exc_info=True)
+                logger.warning(log_message, *log_args, exc_info=True)
                 raise
             except Exception:
                 logger.error(log_message, *log_args, exc_info=True)
diff -r 7ed03689fe01 trytond/trytond/protocols/jsonrpc.py
--- a/trytond/trytond/protocols/jsonrpc.py	Fri May 11 13:52:50 2018 +0200
+++ b/trytond/trytond/protocols/jsonrpc.py	Fri May 11 13:56:27 2018 +0200
@@ -4,6 +4,7 @@
 from decimal import Decimal
 import json
 import base64
+import logging

 from werkzeug.wrappers import Response
 from werkzeug.utils import cached_property
@@ -12,6 +13,8 @@
 from trytond.protocols.wrappers import Request
 from trytond.exceptions import TrytonException

+logger = logging.getLogger(__name__)
+

 class JSONDecoder(object):
