diff -r 7658cd8c2e7f cron.py
--- a/trytond/trytond/modules/company/cron.py	Fri Oct 02 08:56:51 2015 +0200
+++ b/trytond/trytond/modules/company/cron.py	Mon Nov 23 12:35:33 2015 +0100
@@ -23,7 +23,8 @@
         for company in cron.companies:
             User.write([cron.user], {
                     'company': company.id,
-                    'main_company': company.id,
+                    'main_company': (company.parent.id if company.parent
+                        else company.id),
                     })
             with Transaction().set_context(company=company.id):
                 super(Cron, cls)._callback(cron)
