Index: imi/trytond/trytond/modules/purchase_request/purchase_request.py
===================================================================
--- imi.orig/trytond/trytond/modules/purchase_request/purchase_request.py
+++ imi/trytond/trytond/modules/purchase_request/purchase_request.py
@@ -276,6 +276,8 @@ class PurchaseRequest(ModelSQL, ModelVie
         pool = Pool()
         Date = pool.get('ir.date')
         today = Date.today()
+        if not date:
+            date = today
         for product_supplier in product.product_suppliers_used(**pattern):
             supply_date = product_supplier.compute_supply_date(date=today)
             timedelta = date - supply_date
