Index: mai/modules/sale_shipment_grouping/sale.py
===================================================================
--- a/trytond/trytond/modules/sale_shipment_grouping/sale.py
+++ b/trytond/trytond/modules/sale_shipment_grouping/sale.py
@@ -32,7 +32,6 @@ class Sale(metaclass=PoolMeta):
             ('moves.origin', 'like', 'sale.line,%'),
             ('state', 'in', ['draft', 'waiting']),
             ]
-        shipment_domain += self._get_grouped_shipment_planned_date(shipment)
         defaults = Shipment.default_get(self._shipment_grouping_fields,
             with_rec_name=False)
         for field in self._shipment_grouping_fields:
