diff -r 8eadb6e76468 trytond/trytond/modules/sale_number_of_packages/stock.py
--- a/trytond/trytond/modules/sale_number_of_packages/stock.py	Mon Dec 03 16:38:23 2018 +0100
+++ b/trytond/trytond/modules/sale_number_of_packages/stock.py	Wed Dec 12 14:45:49 2018 +0100
@@ -179,7 +179,8 @@

     @classmethod
     def compute_quantities_query(cls, location_ids, with_childs=False,
-            grouping=('product',), grouping_filter=None):
+            grouping=('product',), grouping_filter=None,
+            quantity_field='internal_quantity'):
         pool = Pool()
         Lot = pool.get('stock.lot')
         lot = Lot.__table__()
@@ -187,7 +188,8 @@
         if not Transaction().context.get('normalized_number_of_packages'):
             return super(Move, cls).compute_quantities_query(
                 location_ids, with_childs=with_childs, grouping=grouping,
-                grouping_filter=grouping_filter)
+                grouping_filter=grouping_filter,
+                quantity_field=quantity_field)

         new_grouping = grouping[:]
         new_grouping_filter = (grouping_filter[:] if grouping_filter != None
@@ -199,7 +201,8 @@

         query = super(Move, cls).compute_quantities_query(
             location_ids, with_childs=with_childs, grouping=new_grouping,
-            grouping_filter=new_grouping_filter)
+            grouping_filter=new_grouping_filter,
+            quantity_field=quantity_field)
         if not query:
             return query
