diff -r 7fdc432c6f03 trytond/trytond/modules/sale/sale.py
--- a/trytond/trytond/modules/sale/sale.py	Mon May 06 15:10:35 2019 +0200
+++ b/trytond/trytond/modules/sale/sale.py	Tue Jul 23 16:27:52 2019 +0200
@@ -656,6 +656,9 @@
         default.setdefault('moves', None)
         default.setdefault('shipment_state', 'none')
         default.setdefault('sale_date', None)
+        default.setdefault('untaxed_amount_cache', None)
+        default.setdefault('tax_amount_cache', None)
+        default.setdefault('total_amount_cache', None)
         return super(Sale, cls).copy(sales, default=default)

     def check_for_quotation(self):
