diff -r fb270fa63414 trytond/trytond/modules/sale/sale.py
--- a/trytond/trytond/modules/sale/sale.py	Fri Nov 10 13:30:35 2017 +0100
+++ b/trytond/trytond/modules/sale/sale.py	Fri Nov 10 13:31:38 2017 +0100
@@ -580,17 +580,12 @@
         Search on shipments or returns
         '''
         def method(self, name, clause):
-            nested = clause[0].lstrip(name)
-            if nested:
-                return [('lines.moves.shipment' + nested,)
-                    + tuple(clause[1:3]) + (model_name,) + tuple(clause[3:])]
+            if isinstance(clause[2], basestring):
+                target = 'rec_name'
             else:
-                if isinstance(clause[2], basestring):
-                    target = 'rec_name'
-                else:
-                    target = 'id'
-                return [('lines.moves.shipment.' + target,)
-                    + tuple(clause[1:3]) + (model_name,)]
+                target = 'id'
+            return [('lines.moves.shipment.' + target,) + tuple(clause[1:])
+                + (model_name,)]
         return classmethod(method)
 
     search_shipments = search_shipments_returns('stock.shipment.out')
