diff -r 97d9b0a5b64b trytond/trytond/modules/account_payment_sepa/payment.pypayment.py
--- a/trytond/trytond/modules/account_payment_sepa/payment.py	Mon Jun 05 11:54:05 2017 +0200
+++ b/trytond/trytond/modules/account_payment_sepa/payment.py	Mon Jun 05 11:55:23 2017 +0200
@@ -445,9 +445,9 @@
     def sequence_type(self):
         if self.type == 'one-off':
             return 'OOFF'
-        elif not self.payments:
-            return 'FRST'
-        # TODO manage FNAL
+        # elif not self.payments:
+        #     return 'FRST'
+        # # TODO manage FNAL
         else:
             return 'RCUR'
