diff -r 463332839c4e stock.py
--- a/trytond/trytond/modules/stock_lot/stock.py	Wed May 11 11:26:49 2016 +0200
+++ b/trytond/trytond/modules/stock_lot/stock.py	Fri Jul 08 13:09:53 2016 +0200
@@ -120,16 +120,17 @@
                         round=False)
                     if quantity < out_quantity:
                         outgoing_moves.extend(Move.copy([out_move], default={
-                                    'quantity': out_quantity - quantity,
+                                    'quantity': move.uom.round(
+                                        out_quantity - quantity),
                                     }))
                         Move.write([out_move], {
-                                'quantity': quantity,
+                                'quantity': out_move.uom.round(quantity),
                                 })
                     Move.write([out_move], {
                             'lot': move.lot.id,
                             })
                     quantity -= out_quantity
-                assert quantity <= 0
+                assert quantity <= move.product.default_uom.rounding
 
 
 class ShipmentOutReturn:
