--- a/modules/stock_valued/shipment.py
+++ b/modules/stock_valued/shipment.py
@@ -1,5 +1,5 @@
 #This file is part stock_valued module for Tryton.
-#The COPYRIGHT file at the top level of this repository contains 
+#The COPYRIGHT file at the top level of this repository contains
 #the full copyright notices and license terms.
 
 from trytond.model import fields
@@ -62,7 +62,7 @@ class ShipmentOut:
             context['language'] = self.customer.lang.code
         return context
 
-    def get_untaxed_amount(self, name):
+    def get_untaxed_amount(self):
         '''
         Compute the untaxed amount for each ShipmentOut
         '''
@@ -73,7 +73,7 @@ class ShipmentOut:
         amount = sum((m.amount for m in self.outgoing_moves), Decimal(0))
         return Currency.round(self.company.currency, amount)
 
-    def get_tax_amount(self, name):
+    def get_tax_amount(self):
         '''
         Compute tax amount for each ShipmentOut
         '''
@@ -102,7 +102,7 @@ class ShipmentOut:
                 for key in taxes), Decimal(0))
         return Currency.round(self.company.currency, amount)
 
-    def get_total_amount(self, name):
+    def get_total_amount(self):
         '''
         Return the total amount of each ShipmentOut
         '''
