diff -r c8f4cfc59666 trytond/trytond/modules/stock_lot_deactivatable/stock.py
--- a/trytond/trytond/modules/stock_lot_deactivatable/stock.py	Wed Dec 12 15:46:19 2018 +0100
+++ b/trytond/trytond/modules/stock_lot_deactivatable/stock.py	Wed Dec 12 15:50:28 2018 +0100
@@ -82,13 +82,14 @@

     @classmethod
     def compute_quantities_query(cls, location_ids, with_childs=False,
-            grouping=('product',), grouping_filter=None):
+            grouping=('product',), grouping_filter=None,
+            quantity_field='internal_quantity'):
         pool = Pool()
         Lot = pool.get('stock.lot')
         Period = pool.get('stock.period')
         query = super(Move, cls).compute_quantities_query(
             location_ids, with_childs=with_childs, grouping=grouping,
-            grouping_filter=grouping_filter)
+            grouping_filter=grouping_filter, quantity_field=quantity_field)

         if not query or 'lot' not in grouping:
             return query
