diff --git a/shipment.py b/shipment.py
index 6ef998f..bd94381 100644
--- a/shipment.py
+++ b/shipment.py
@@ -132,16 +132,18 @@ class ShipmentInternal(ModelSQL, ModelView):
                     state='request',
                     )
                 shipment_moves = []
+
                 for move in moves:
                     product_id, qty = move
                     product = id2product.setdefault(
                         product_id, Product(product_id))
+                    digits = product.default_uom.digits
                     shipment_moves.append(Move(
                             from_location=from_location,
                             to_location=to_location,
                             planned_date=date,
                             product=product,
-                            quantity=qty,
+                            quantity=round(qty, digits),
                             uom=product.default_uom,
                             company=user_record.company,
                             ))
