diff -r d3533b14b811 inventory.py
--- a/trytond/trytond/modules/stock/inventory.py	Thu Mar 01 23:07:13 2018 +0100
+++ b/trytond/trytond/modules/stock/inventory.py	Fri Mar 02 03:03:20 2018 +0100
@@ -218,15 +218,13 @@
 
             # Index some data
             product2type = {}
-            product2consumable = {}
             for product in Product.browse([line[1] for line in pbl]):
                 product2type[product.id] = product.type
-                product2consumable[product.id] = product.consumable
 
             # Update existing lines
             for line in inventory.lines:
-                if not (line.product.type == 'goods'
-                        and not line.product.consumable):
+                if not (line.product.active and
+                        line.product.type == 'goods'):
                     Line.delete([line])
                     continue
 
@@ -244,8 +242,7 @@
             # Create lines if needed
             for key, quantity in pbl.iteritems():
                 product_id = key[grouping.index('product') + 1]
-                if (product2type[product_id] != 'goods'
-                        or product2consumable[product_id]):
+                if product2type[product_id] != 'goods':
                     continue
                 if not quantity:
                     continue
@@ -265,7 +262,6 @@
     product = fields.Many2One('product.product', 'Product', required=True,
         domain=[
             ('type', '=', 'goods'),
-            ('consumable', '=', False),
             ])
     uom = fields.Function(fields.Many2One('product.uom', 'UOM'), 'get_uom')
     unit_digits = fields.Function(fields.Integer('Unit Digits'),
