diff -r 36fa08276f66 trytond/trytond/modules/stock_consignment/stock.py
--- a/trytond/trytond/modules/stock_consignment/stock.py Thu Mar 01 23:04:06 2018 +0100
+++ b/trytond/trytond/modules/stock_consignment/stock.py Wed May 30 11:40:24 2018 +0200
@@ -226,7 +226,7 @@
     @classmethod
     @ModelView.button
     @Workflow.transition('assigned')
-    @set_origin_consignment
+    @unset_origin_consignment
     def assign(cls, moves):
         super(Move, cls).assign(moves)
 
--- a/trytond/trytond/modules/stock/move.py	2018-06-06 10:34:21.750490990 +0200
+++ b/trytond/trytond/modules/stock/move.py	2018-06-06 10:35:03.254705489 +0200
@@ -621,7 +621,7 @@
     @ModelView.button
     @Workflow.transition('assigned')
     def assign(cls, moves):
-        cls.check_origin(moves)
+        pass
 
     @classmethod
     @ModelView.button
