--- a/modules/stock_valued_discount/shipment.py
+++ b/modules/stock_valued_discount/shipment.py
@@ -1,5 +1,5 @@
 #This file is part stock_valued_discount module for Tryton.
-#The COPYRIGHT file at the top level of this repository contains 
+#The COPYRIGHT file at the top level of this repository contains
 #the full copyright notices and license terms.
 from trytond.transaction import Transaction
 from trytond.pool import Pool, PoolMeta
@@ -13,7 +13,7 @@ class ShipmentOut:
     "Customer Shipment"
     __name__ = 'stock.shipment.out'
 
-    def get_tax_amount(self, name):
+    def get_tax_amount(self):
         '''
         Compute tax amount for each ShipmentOut
         '''
