diff --git a/tax.py b/tax.py
--- a/trytond/trytond/modules/account/tax.py	Thu Gen 20 10:25:13 2020 +0200
+++ b/trytond/trytond/modules/account/tax.py	Thu Gen 20 10:32:48 2020 +0200
@@ -893,7 +893,8 @@ class Tax(sequence_ordered(), ModelSQL, ModelView, DeactivableMixin):
                 if tax.type != 'none':
                     value = tax._process_tax(price_unit)
                     res.append(value)
-                    if tax.update_unit_price:
+                    if (tax.update_unit_price or
+                            tax.parent and tax.parent.update_unit_price):
                         unit_price_variation += value['amount']
                 if len(tax.childs):
                     res.extend(cls._unit_compute(tax.childs, price_unit, date))
