diff -r bbae9fd78f9e trytond/backend/postgresql/database.py
--- a/trytond/trytond/backend/postgresql/database.py	Thu Mar 28 19:22:49 2019 +0100
+++ b/trytond/trytond/backend/postgresql/database.py	Mon Apr 01 14:39:22 2019 +0200
@@ -291,7 +291,7 @@
     def _test(cls, connection, hostname=None):
         cursor = connection.cursor()
         tables = ('ir_model', 'ir_model_field', 'ir_ui_view', 'ir_ui_menu',
-            'res_user', 'res_group', 'ir_module', 'ir_module_dependency',
+            'res_user', 'res_group',
             'ir_translation', 'ir_lang', 'ir_configuration')
         cursor.execute('SELECT table_name FROM information_schema.tables '
             'WHERE table_name IN %s', (tables,))
