diff -r 6c0d835dd918 trytond/trytond/model/workflow.py
--- a/trytond/trytond/model/workflow.py	Wed May 11 10:40:32 2016 +0200
+++ b/trytond/trytond/model/workflow.py	Mon May 30 16:14:16 2016 +0200
@@ -21,7 +21,7 @@
             @wraps(func)
             def wrapper(cls, records, *args, **kwargs):
                 filtered = []
-                to_update = OrderedDict()
+                to_check = []
 
                 for record in records:
                     current_state = getattr(record, cls._transition_state)
@@ -29,17 +29,19 @@
                     if transition in cls._transitions:
                         filtered.append(record)
                         if current_state != state:
-                            to_update[record] = current_state
+                            to_check.append((record, current_state))
 
                 result = func(cls, filtered, *args, **kwargs)
-                if to_update:
-                    for record in to_update.keys():
+                if to_check:
+                    to_update = []
+                    for record, previous_state in to_check:
                         current_state = getattr(record, cls._transition_state)
-                        if current_state != to_update[record]:
-                            del to_update[record]
-                    cls.write(list(to_update), {
-                            cls._transition_state: state,
-                            })
+                        if current_state == previous_state:
+                            to_update.append(record)
+                    if to_update:
+                        cls.write(to_update, {
+                                cls._transition_state: state,
+                                })
                 return result
             return wrapper
         return check_transition
